server_cached_router.go

```diff
@@ -0,0 +1,247 @@
+		record.Val.Addrs = r.withAddrsFromCache(addrQueryOriginPeers, *record.Val.ID, record.Val.Addrs)
+		return record
+	}), nil
+}
+
+// withAddrsFromCache returns the best list of addrs for specified [peer.ID].
+// It will consult cache ONLY if the addrs slice passed to it is empty.
+func (r cachedRouter) withAddrsFromCache(queryOrigin string, pid peer.ID, addrs []types.Multiaddr) []types.Multiaddr {
+	// skip cache if we already have addrs
+	if len(addrs) > 0 {
+		peerAddrLookups.WithLabelValues(addrCacheStateUnused, queryOrigin).Inc()
+		return addrs
+	}
+
+	cachedAddrs := r.cachedAddrBook.GetCachedAddrs(pid)
+	if len(cachedAddrs) > 0 {
+		logger.Debugw("found cached addresses", "peer", pid, "cachedAddrs", cachedAddrs)
+		peerAddrLookups.WithLabelValues(addrCacheStateHit, queryOrigin).Inc()
+		return cachedAddrs
+	} else {
+		// Cache miss. Queue peer for lookup.
+		peerAddrLookups.WithLabelValues(addrCacheStateMiss, queryOrigin).Inc()
+		return nil
+	}
+}
+
+var _ iter.ResultIter[types.Record] = &cacheFallbackIter{}
+
+type cacheFallbackIter struct {
+	sourceIter      iter.ResultIter[types.Record]
+	current         iter.Result[types.Record]
+	findPeersResult chan types.PeerRecord
+	router          cachedRouter
+	ctx             context.Context
+	cancel          context.CancelFunc
+	ongoingLookups  atomic.Int32
+}
+
+// NewCacheFallbackIter is a wrapper around a results iterator that will resolve peers with no addresses from cache and if no cached addresses, will look them up via FindPeers.
+// It's a bit complex because it ensures we continue iterating without blocking on the FindPeers call.
+func NewCacheFallbackIter(sourceIter iter.ResultIter[types.Record], router cachedRouter, ctx context.Context) *cacheFallbackIter {
+	ctx, cancel := context.WithCancel(ctx)
+	return &cacheFallbackIter{
+		sourceIter:      sourceIter,
+		router:          router,
+		ctx:             ctx,
+		cancel:          cancel,
+		findPeersResult: make(chan types.PeerRecord),
+		ongoingLookups:  atomic.Int32{},
+	}
+}
+
+func (it *cacheFallbackIter) Next() bool {
+	// Try to get the next value from the source iterator first
+	if it.sourceIter.Next() {
+		val := it.sourceIter.Val()
+		handleRecord := func(id *peer.ID, record *types.PeerRecord) bool {
+			record.Addrs = it.router.withAddrsFromCache(addrQueryOriginProviders, *id, record.Addrs)
+			if record.Addrs != nil {
+				it.current = iter.Result[types.Record]{Val: record}
+				return true
+			}
+			logger.Infow("no cached addresses found in cacheFallbackIter, dispatching find peers", "peer", id)
+
+			if it.router.cachedAddrBook.ShouldProbePeer(*id) {
+				it.ongoingLookups.Add(1) // important to increment before dispatchFindPeer
+				// If a record has no addrs, we dispatch a lookup to find addresses
+				go it.dispatchFindPeer(*record)
+
+				return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
```

Comment:
Will this exit the loop if `it.Next()` returns false? Do you need to wait for the find peer call dispatched just before to end?

Reply:
This is probably my unfamiliarity with the construct. Please ignore if it doesn't make sense.

Reply:
Great question. The recursive `it.Next()` call will block on ongoing lookups (if the underlying iterator is done but we have an ongoing dispatched find peer), thereby ensuring that we return the value based on whatever the result of the dispatched lookup.

Reply:
Ah, my bad. For some reason I read that as source.Next.